pytorch_blade/src/torch-mlir/lib/Conversion/MhloPasses.cpp

```diff
@@ -11,19 +11,157 @@
+    pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
+    // Inline global slots, which for most inference scenarios deletes them.
+    // This also exposes more information to intraprocedural transformations
+    // below like MaximizeValueSemantics and RefineTypes.
+    // OPT-ONLY: Don't rely on this pass to "lower" global slots by deleting.
+    // Also don't rely on this pass to expose constants into the program to
+    // simplify handling of "optional".
+    pm.addPass(createInlineGlobalSlotsPass());
+  }
+
+  // Reduce variants of ops to a smaller set of primitives.
+  pm.addNestedPass<func::FuncOp>(createReduceOpVariantsPass());
+
+  if (options.optimize) {
+    // OPT-ONLY: Right now we rely on this to eliminate certain branches that
+    // guard unreachable code that backends can't handle yet, such as lists,
+    // RaiseException, unimplemented tensor ops, and only-used-in-training
+    // operations on `torch.global_slot`'s.
+    pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
+    // OPT-ONLY: We may have deleted some `torch.global_slot.get` /
+    // `torch.global_slot.get` ops, which may have left more
+    // `torch.global_slot`'s unused.
+    pm.addPass(createSymbolDCEPass());
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Lowering to ranked !torch.vtensors of known dtype.
+  //===--------------------------------------------------------------------===//
+
+  // Convert the bulk of non-ABI-visible !torch.tensor's to !torch.vtensor's.
+  pm.addNestedPass<func::FuncOp>(Torch::createMaximizeValueSemanticsPass());
+
+  // Do shape refinement.
+  // This must be run before RefineTypes (which primarily does dtype inference),
+  // because Torch type promotion rules actually depend on the shape of the
+  // operand.
+  // createTorchShapeRefinementPipeline(pm, options);
+  // Refine types in the program, which mainly means inferring dtypes of ops.
+  // pm.addNestedPass<func::FuncOp>(Torch::createRefineTypesPass());
+
+  // Propagate to ABI return types the shape/dtype information discovered by
+  // the previous pass. Doing this is ABI-compatible for our backends.
+  pm.addPass(Torch::createRefinePublicReturnPass());
+
+  if (options.optimize) {
+    // This can fold away some branches given the information got from
+    // RefineTypes before doing maximize value sematics which only works with
+    // basic blocks.
+    pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
+  }
+
+  if (options.optimize) {
+    // All the type refinement we've done above has exposed new information
+    // that allows folding away more stuff.
+    // OPT-ONLY: Right now we rely on this to eliminate certain
+    // branches that guard unreachable code that backends can't handle yet, such
+    // as lists, RaiseException, unimplemented aten ops, and
+    // only-used-in-training operations on `torch.global_slot`'s.
+    pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
```

Comment:
duplicate code? line 151 and line 158

Reply:
just copy from the community, but I think we can remove the duplicate pass in our pass pipeline. Maybe we will do that in the next PR because the heave CI ....